libs/utils/trace_analysis.py

```diff
@@ -1009,52 +888,107 @@ def getClusterFrequencyResidency(self, cluster):
-            'ACTIVE Time' : time_intervals,
-            'Frequency [MHz]' : [f/1000 for f in frequencies]
+        :param cluster: list of CPU IDs belonging to the cluster
+        :type cluster: list(int)
+        """
+
+        if not self.trace.hasEvents('cpu_frequency'):
+            logging.warn('Events [cpu_frequency] not found, '\
+                         'plot DISABLED!')
+            return None
+        if not self.trace.hasEvents('cpu_idle'):
+            logging.warn('Events [cpu_idle] not found, '\
+                         'plot DISABLED!')
+            return None
+
+        freq_dfr = self.trace.df('cpu_frequency')
+        cluster_freqs = freq_dfr[freq_dfr.cpu.isin(cluster)]
+        idle_dfr = self.trace.df('cpu_idle')
+
+        ### Computer TOTAL Time ###
+        time_intervals = cluster_freqs.index[1:] - cluster_freqs.index[:-1]
+        total_time = pd.DataFrame({
+            'TOTAL Time' : time_intervals,
+            'Frequency [MHz]' : [
+                f/1000 for f in cluster_freqs.iloc[:-1].frequency
+            ]
         })
-        # Sum the time intervals with same frequncy
-        active_time = active_time.groupby(['Frequency [MHz]']).sum()
+        total_time = total_time.groupby(['Frequency [MHz]']).sum()
+
+        ### Compute ACTIVE Time ###
+        cluster_active = self._getClusterActiveSignal(cluster)
+
+        # In order to compute the active time spent at each frequency we
+        # multiply 2 square waves:
+        # - cluster_active, a square of the form:
+        #           cluster_active[t] = 1 if there is a task running in the
+        #                               cluster at time t
+        #           cluster_active[t] = 0 otherwise
+        # - freq_active, square wave of the form:
+        #           freq_active[t] = 1 if at time t the frequency is f
+        #           freq_active[t] = 0 otherwise
+        available_freqs = sorted(cluster_freqs.frequency.unique())
+        new_idx = sorted(cluster_freqs.index.tolist() + \
+                         cluster_active.index.tolist())
+        cluster_freqs = cluster_freqs.reindex(
+            new_idx, method='ffill'
+        ).fillna(method='bfill')
+        cluster_active = cluster_active.reindex(new_idx, method='ffill')
+        nonidle_time = []
+        for f in available_freqs:
+            # Build square wave for the current frequency such that:
+            freq_active = cluster_freqs.frequency.apply(
+                lambda x: 1 if x == f else 0
+            )
+            active_t = cluster_active * freq_active
+            # Sum up the time intervals by computing the integral of the square
+            # wave
+            nonidle_time.append(self._integrate_square_wave(active_t))
+
+        active_time = pd.DataFrame({"ACTIVE Time" : nonidle_time},
+                                   index=[f/1000 for f in available_freqs])
+        active_time.index.name = 'Frequency [MHz]'
         return total_time, active_time
```

Comment:
Do you think it is possible to use a memoize decorator to save both active and total time once computed by this function: https://wiki.python.org/moin/PythonDecoratorLibrary#Memoize

This would allow to call the same function multiple times without redoing the computations.

Does it make sense?
I was thinking that for example this can be useful when we want to figure out the x_max across multiple clusters.
In this case we likely need to call the function one time for each cluster to compute the x_max and than again to plot the values.